File: mobile/src/services/geofenceChecker.js
// Point-in-polygon algorithm to check if GPS coordinate is inside a geofence
const isPointInPolygon = (point, polygon) => {
    const { lat, lng } = point;
    const coordinates = polygon.coordinates[0]; // GeoJSON polygon format

    let inside = false;
    for (let i = 0, j = coordinates.length - 1; i < coordinates.length; j = i++) {
        const xi = coordinates[i][0]; // longitude
        const yi = coordinates[i][1]; // latitude
        const xj = coordinates[j][0];
        const yj = coordinates[j][1];

        const intersect = ((yi > lat) !== (yj > lat)) &&
            (lng < (xj - xi) * (lat - yi) / (yj - yi) + xi);

        if (intersect) inside = !inside;
    }

    return inside;
};

// Apply time-based rules if present
const getEffectiveLimit = (zone) => {
    const now = new Date();
    const currentHour = now.getHours();

    if (!zone.time_rules || Object.keys(zone.time_rules).length === 0) {
        return zone.speed_limit;
    }

    // Check each time period
    for (const [period, rules] of Object.entries(zone.time_rules)) {
        if (rules.start && rules.end && rules.limit) {
            const [startH] = rules.start.split(':').map(Number);
            const [endH] = rules.end.split(':').map(Number);

            const inPeriod = startH < endH
                ? (currentHour >= startH && currentHour < endH)
                : (currentHour >= startH || currentHour < endH);

            if (inPeriod) {
                return Math.min(zone.speed_limit, rules.limit);
            }
        }
    }

    return zone.speed_limit;
};

// Get applicable speed limit for current position
export const getSpeedLimitForPosition = (lat, lng, geofences, loadType = 'empty') => {
    // Load-type base limits (from blueprint)
    const loadLimits = {
        empty: 50,
        partial: 50,
        full: 40,
        hazardous: 30,
    };

    const baseLoadLimit = loadLimits[loadType] || 50;
    let applicableLimit = 50; // Plant default
    let currentZone = null;

    // Check each geofence
    for (const zone of geofences) {
        if (isPointInPolygon({ lat, lng }, zone.polygon)) {
            const zoneLimit = getEffectiveLimit(zone);

            // Hazardous load special rules
            const adjustedLimit = loadType === 'hazardous' && zone.zone_type === 'pedestrian'
                ? Math.min(zoneLimit, 10)
                : zoneLimit;

            // Take most restrictive limit
            if (adjustedLimit < applicableLimit) {
                applicableLimit = adjustedLimit;
                currentZone = zone;
            }
        }
    }

    // Apply load limit
    const finalLimit = Math.min(applicableLimit, baseLoadLimit);

    return {
        limit: finalLimit,
        zone: currentZone,
        insideZone: !!currentZone,
    };
};

File: backend/services/geofence_checker.py
"""
Geofence Checker Service
Given a GPS coordinate, returns the applicable speed limit.
This is called on EVERY position update — must be fast.
"""
# pyrefly: ignore [missing-import]
from sqlalchemy.ext.asyncio import AsyncSession
# pyrefly: ignore [missing-import] 
from sqlalchemy import text
from datetime import datetime
import json


async def get_speed_limit_for_position(
    lat: float,
    lng: float,
    db: AsyncSession,
    load_type: str = "empty"
) -> dict:
    """
    Returns the applicable speed limit for a given GPS position.
    
    Logic:
    1. Query PostGIS for any active geofence polygon containing this point
    2. Apply time-based rules if present  
    3. Apply load-type adjustments
    4. Return the MINIMUM speed limit (most restrictive applies)
    
    The ST_Within query uses the spatial GIST index and is very fast.
    """
    
    result = await db.execute(
        text("""
            SELECT id, name, speed_limit, zone_type, time_rules
            FROM geofences
            WHERE active = true
            AND ST_Within(
                ST_SetSRID(ST_MakePoint(:lng, :lat), 4326),
                polygon
            )
            ORDER BY speed_limit ASC
            LIMIT 1
        """),
        {"lat": lat, "lng": lng}
    )
    zone = result.fetchone()
    
    # Default plant-wide limit
    default_limit = 50
    
    # Load-type overrides (from blueprint section on load-aware limits)
    load_limits = {
        "empty": 50,
        "partial": 50,
        "full": 40,
        "hazardous": 30
    }
    load_limit = load_limits.get(load_type, 50)
    
    if zone:
        zone_limit = zone.speed_limit
        
        # Check time-based rules
        time_rules = zone.time_rules if isinstance(zone.time_rules, dict) else json.loads(zone.time_rules or "{}")
        if time_rules:
            current_hour = datetime.now().hour
            for period, rules in time_rules.items():
                if "start" in rules and "end" in rules:
                    start_h = int(rules["start"].split(":")[0])
                    end_h = int(rules["end"].split(":")[0])
                    in_period = (
                        (start_h <= current_hour < end_h) if start_h < end_h
                        else (current_hour >= start_h or current_hour < end_h)
                    )
                    if in_period and "limit" in rules:
                        zone_limit = min(zone_limit, rules["limit"])
        
        # Apply load-type adjustment within zone
        if load_type == "hazardous":
            zone_limit = min(zone_limit, 10)  # Max 10 km/h for hazardous in any zone
        
        # Take the most restrictive limit
        effective_limit = min(zone_limit, load_limit)
        
        return {
            "speed_limit": effective_limit,
            "zone_name": zone.name,
            "zone_id": str(zone.id),
            "zone_type": zone.zone_type,
            "inside_zone": True
        }
    
    # No zone found — apply plant default and load limit
    return {
        "speed_limit": min(default_limit, load_limit),
        "zone_name": "Plant Road",
        "zone_id": None,
        "zone_type": "main_road",
        "inside_zone": False
    }